google-cloud-firestore/src/main/java/com/google/cloud/firestore/Query.java

```diff
@@ -459,35 +463,53 @@ private static boolean isUnaryComparison(@Nullable Object value) {
     return value == null || value.equals(Double.NaN) || value.equals(Float.NaN);
   }
 
-  /** Computes the backend ordering semantics for DocumentSnapshot cursors. */
-  private ImmutableList<FieldOrder> createImplicitOrderBy() {
-    List<FieldOrder> implicitOrders = new ArrayList<>(options.getFieldOrders());
+  /** Returns the sorted set of inequality filter fields used in this query. */
+  private SortedSet<FieldPath> getInequalityFilterFields() {
+    SortedSet<FieldPath> result = new TreeSet<>();
 
-    // If no explicit ordering is specified, use the first inequality to define an implicit order.
-    if (implicitOrders.isEmpty()) {
-      for (FilterInternal filter : options.getFilters()) {
-        FieldReference fieldReference = filter.getFirstInequalityField();
-        if (fieldReference != null) {
-          implicitOrders.add(new FieldOrder(fieldReference, Direction.ASCENDING));
-          break;
+    for (FilterInternal filter : options.getFilters()) {
+      for (FieldFilterInternal subFilter : filter.getFlattenedFilters()) {
+        if (subFilter.isInequalityFilter()) {
+          result.add(FieldPath.fromServerFormat(subFilter.fieldReference.getFieldPath()));
         }
       }
     }
 
-    boolean hasDocumentId = false;
-    for (FieldOrder fieldOrder : implicitOrders) {
-      if (FieldPath.isDocumentId(fieldOrder.fieldReference.getFieldPath())) {
-        hasDocumentId = true;
+    return result;
+  }
+
+  /** Computes the backend ordering semantics for DocumentSnapshot cursors. */
+  private ImmutableList<FieldOrder> createImplicitOrderBy() {
+    // Any explicit order by fields should be added as is.
+    List<FieldOrder> implicitOrders = new ArrayList<>(options.getFieldOrders());
```

Comment:
I think you should rename `implicitOrders` to `result`. Because it's not just the implicit orderBys. It gets initialized with explicit orderBys, and others get added to it.